binaries/cuprated/src/killswitch.rs

```diff
@@ -0,0 +1,82 @@
+//! Killswitch.
+//!
+//! This module implements code for shutting down `cuprated`
+//! after a certain timestamp has passed.
+//!
+//! The reasoning is twofold:
+//! 1. Limiting the effects of any network errors
+//!    caused by a faulty `cuprated`.
+//! 2. To enforce users to update `alpha` builds,
+//!    if they choose to run them.
+//!
+//! This behavior is limited to an alpha build;
+//! this module will be removed after a stable v1 release.
+
+use std::{process::exit, time::Duration};
+
+use cuprate_helper::time::current_unix_timestamp;
+
+/// Assert that this is not a v1 release and an alpha release.
+pub const _: () = {
+    const_format::assertcp_ne!(
+        crate::constants::MAJOR_VERSION,
+        "1",
+        "`cuprated` major version is 1, killswitch module should be deleted."
+    );
+};
+
+/// The killswitch activates if the current timestamp is ahead of this timestamp.
+///
+/// Sat Mar 01 2025 05:00:00 GMT+0000
+pub const KILLSWITCH_ACTIVATION_TIMESTAMP: u64 = 1740805200;
+
+/// Check if the system clock is past a certain timestamp,
+/// if so, exit the entire program.
+pub fn killswitch() {
+    /// A timestamp known to have been passed.
+    ///
+    /// This is an arbitrary timestamp used for
+    /// sanity checking the system's clock to make
+    /// sure it is not overly behind.
+    ///
+    /// Fri Jan 17 2025 14:19:10 GMT+0000
+    const SYSTEM_CLOCK_SANITY_TIMESTAMP: u64 = 1737123550;
+
+    /// Print a generic killswitch message.
+    fn print_killswitch_msg(msg: String) {
+        eprintln!("killswitch: {msg}. `cuprated` will now exit. For more details on why this exists, see: <https://github.com/Cuprate/cuprate/blob/main/binaries/cuprated/src/killswitch.rs>.");
+    }
+
+    let current_ts = current_unix_timestamp();
+
+    if current_ts < SYSTEM_CLOCK_SANITY_TIMESTAMP {
+        print_killswitch_msg(format!("The system clock is too far behind (current_ts: {current_ts}) and is not reliable to use"));
+        exit(66);
+    }
```

Comment:
6/6

This guards against a system with a messed up system clock from evading the killswitch.

```bash
timedatectl set-ntp false
timedatectl set-time 2010-03-06
date -u
```

> Sat Mar  6 05:00:01 AM UTC 2010

```bash
cargo run --bin cuprated
```

> killswitch: The system clock is too far behind and is not reliable to use (current_ts: 1267851602, killswitch_activation_timestamp: 1740805200). `cuprated` will now exit. For more details on why this exists, see: <https://github.com/Cuprate/cuprate/blob/main/binaries/cuprated/src/killswitch.rs>.